sycl/doc/CompilerAndRuntimeDesign.md

```diff
@@ -787,18 +787,60 @@ The SPIR-V specific functions are implemented in for the SYCL host device here:
 
 ### Address spaces handling
 
-SYCL 1.2.1 language defines several address spaces where data can reside, the
-same as OpenCL - global, local, private and constant. From the spec: "In OpenCL
-C, these address spaces are manually specified using OpenCL-specific keywords.
-In SYCL, the device compiler is expected to auto-deduce the address space for
-pointers in common situations of pointer usage. However, there are situations
-where auto-deduction is not possible".
-
-We believe that requirement for the compiler to automatically deduce address
-spaces is too strong. Instead the following approach will be implemented in the
-compiler:
+SYCL specification uses C++ classes to represent pointers to disjoint memory
+regions on an accelerator to enable compilation with standard C++ toolchain and
+SYCL compiler toolchain.
+
+For instance:
+
+``` C++
+// check that SYCL mode is ON and we can use non-standard decorations
+#if defined(__SYCL_DEVICE_ONLY__)
+// GPU/accelerator implementation
+template <typename T, address_space AS> class multi_ptr {
+  // DecoratedType<T, global>::type == "__attribute__((opencl_global)) T"
+  // See sycl/include/CL/sycl/access/access.hpp for more details
+  using pointer_t = typename DecoratedType<T, AS>::type *;
+ 
+  pointer_t m_Pointer;
+  public:
+  pointer_t get() { return m_Pointer; }
```

Comment:
We use decorated pointer with get() method. There is also subscript operator. It  returns reference which can be also decorated. In this case the code like below we will have decorated pointer which might be helpful in some cases:
```
        queue.submit([&](cl::sycl::handler& cgh)
        {
            auto acc = buf.get_access<cl::sycl::access::mode::read_write>(cgh);
            cgh.parallel_for<class test>(cl::sycl::nd_range<1>(N, 16), [=](cl::sycl::nd_item<1> it)
            {
                auto x = &acc[i];
             }
         }
```
C++ preserve decoration during getting pointer from reference. So the only change required is to return decorate reference from subscript operator.

Reply:
I'm not quite sure what are you proposing here. Do you mean we should add another member with a reference to decorated type? It doesn't seem to expose a new use case for address space attribute. Right?

What might be worth adding is `multi_ptr` subscript operator, which drops AS decoration in DPC++ implementation to demonstrate the case when AS-decorated pointer is casted to undecorated/default pointer.